### ISSUE TITLE
Stop using bakefile 0.x

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#23249]
Unfortunately [this thread](https://groups.google.com/g/wx-dev/c/XSFIlZ85fG8/m/LfOJ249GAgAJ) didn't really result in any confusion, but we still need to stop depending on unmaintained and obsolete bakefile.

Things currently preventing us from doing it:

- [ ] `Makefile.in` for building the library itself under Unix. In the worst case, we can maintain this one manually. Maybe we could try to use automake to facilitate this, but I'm not sure how well does it work without libtool and I am almost sure that we don't want to use libtool.
- [ ] Samples (and utils, tests etc) makefiles. There are too many of them to maintain them by hand, so we need some tool to help us. [Premake](https://premake.github.io/) could be a good choice, but it would be really great to have some help from somebody already familiar with it. [Xmake](https://xmake.io/) is another possibility. Finally, if all else fails, we can write our own ad hoc tool, but I'd rather avoid this.

**Comment by vadz** [comment]
As written in [this post](https://groups.google.com/g/wx-dev/c/XSFIlZ85fG8/m/CSquMRrCCQAJ) things I'd like to do if we decide to use Premake:

- [ ] Produce MSVS version-independent project files, i.e. avoid defining `PlatformToolset` in the generated files and import `build/msw/wx_config.props` instead, to have just a single `.vcxproj` that can be used with any currently supported MSVS version.
- [ ] Find a way to centralize the common parts of all samples files and have a common module for them.
- [ ] Add dependencies on wx libraries to the sample projects.

As mentioned before, any help from people familiar with Premake/Lua (@pkulchenko ?) would be very welcome, as I have a feeling that all this should be quite possible to do, but don't have any idea where to start. TIA!

**Comment by vadz** [comment]
I've done one thing to facilitate migrating from bakefile: in #24937 I've added MSVS project files for all the samples (and we already had them for the tests). This leaves demos and utils, but

1. I wonder if demos shouldn't be just dropped, they're completely unmaintained and really don't look as impressive in 2024 as they did in 1994.
2. Among utils, only wxrc is really useful and I think we could add a manually created project file for it too, it doesn't change very often.

If/when we do this, we could remove all `makefile.vc` files. This still leaves `makefile.unx` and `makefile.gcc`, which I'm also tempted to just drop, and `Makefile.in` ones — which we definitely need to preserve and find some way to deal with updating them. Unfortunately I still don't see any good way forward here, my attempts to use Premake have failed...

**Comment by eranif** [comment]
We already have good `CMake` (this is how I build wxWidgets on Windows) - why add another tool like `Premake` or `Xmake` (?)
BTW, (never used it) but `CMake` can generate MSVC projects as well

**Comment by vadz** [comment]
I don't want to force people to install CMake and as long as we have another build system (configure-based) anyhow, it should be possible to build the samples etc using it too.

**Comment by eranif** [comment]
I am guessing that I will need to install `premake`? or `xmake`? 
Beside, these days, `CMake` is available on platforms that wxWidgets is supported

**Comment by vadz** [comment]
Both these tools can generate make/project files usable without them (just as bakefile did). But we're very far from using either of them anyhow.

**Comment by vadz** [comment]
Unfortunately this is not going to happen before 3.3.0, we're still stuck with bakefile for the time being.

**Comment by vadz** [comment]
It would be great if somebody could contribute at least a skeleton XMake-based system that would be capable of generating XCode projects.

**Comment by Brickmaster1** [comment]
> It would be great if somebody could contribute at least a skeleton XMake-based system that would be capable of generating XCode projects.

Oh did you try and not have success with XCode??

**Comment by Brickmaster1** [comment]
Personally I am really working to try and get the current CMake-based xmake package for wxWidgets to build under the Zig cross compiling toolchain for C++. It is a tall task and I have not yet been successful, but if I can get it to work it means you only need one machine, and you can build every version of the package/library.

**Comment by Brickmaster1** [comment]
XMake is mostly platform agnostic.

**Comment by Brickmaster1** [comment]
> It would be great if somebody could contribute at least a skeleton XMake-based system that would be capable of generating XCode projects.

Just realized about the previous comment, I suppose you wouldn't know since there were never any basic xmake attempts made, but xmake supports XCode as a toolchain, you would need xmake to initially make commits to the project related to build system matters, but it supports generating XCode project files behind the scenes through first making cmakelists and then having cmake generate XCode project. It can also generate makefiles, cmakelists, and build.ninja, and many more. It is as expansive as you want it to be, but in my experience using the xmake tools itself is the easiest and smoothest way to get things done with the most support. There is nothing saying you couldn't just provide generated versions of these files with each important commit, it will just increase the build time some, so maybe be conservative and have it be a toggle during build by default. XMake also makes including libraries much more flexible, and makes it much more viable if you wanted to, for example, build X11 or Wayland on windows to then compile the program for Linux. MacOS is kinda tough because you will need a sysroot for the frameworks, but possible if you have a way of obtaining one (I know there are a couple ways floating around on github).

**Comment by vadz** [comment]
To be clear, what I'm really interested in is a bakefile-like generator for the "native" project files for MSVS and XCode. We already can build using these toolchains with CMake, but it's not that great to work with CMake-generated projects, so something generating more "human-like" projects would be a plus.

**Comment by generic-pers0n** [comment]
I'd like to throw in my 2 cents here. Before that, I must make a disclaimer that I'm not familiar with the two, but if needed, I'm up for learning more about them. Maybe I can figure out how to make a new Premake or Xmake build system for wxWidgets if desired, but we'll see.

From what I understand, it appears that Premake might be the most suitable option based on what I've read from this issue and others. It seems like the goal is to keep a build system that generates usable build files where users can use their favorite build tool (`make`, Xcode, Visual Studio, etc.) without having to have Premake installed or anything. In a way, this is similar to CMake except the actual generated files are usable (as, from my understanding, CMake-generated projects are not meant to be used outside the build process).

Xmake, on the other hand, looks pretty neat! However, I understand it to be yet another actual build system rather than just a meta-buildsystem, although "it can generate project files like CMake or Meson" (according to [it's introduction](https://xmake.io/guide/introduction.html)). Knowing how CMake-generated build files aren't supposed to be used outside the build process, this sounds like those build files would also end up the same way.

**TL;DR:** I think Premake would be the most suitable option for replacing bakefiles.

**Comment by vadz** [comment]
I also thought Premake would be the best solution, but after trying to use it for the samples, I realized that a lot of work would be needed to reproduce what we have currently. I don't remember all the details any more, but I think there was a problem with being able to use either static or dynamic wx libraries and factoring out setting common to all samples was also not that easy to do. So while I still think that it should be possible to use Premake for us, it seems very unlikely that I ever find enough time to do it myself, i.e. it would need to be done by somebody else or not done at all.

Xmake looked promising but I don't know if we can really commit to it. And, of course, somebody still needs to write all `xmake.lua` files...

To be honest, the most likely scenario seems to be that we just abandon all generated build files, i.e. we'll have only

1. CMake
2. MSVS project files
3. `Makefile.in` used with `configure`

This is not ideal but I don't see us releasing 3.4 still using bakefile 0.x and I don't see any other realistic solutions :-(

**Comment by generic-pers0n** [comment]
> To be honest, the most likely scenario seems to be that we just abandon all generated build files, i.e. we'll have only
> 
>     1. CMake
> 
>     2. MSVS project files
> 
>     3. `Makefile.in` used with `configure`
> 
> 
> This is not ideal but I don't see us releasing 3.4 still using bakefile 0.x and I don't see any other realistic solutions :-(

I'll be completely honest, I'm for using CMake, mainly because you have the flexibility of doing what you please with it and also because it's really the only build system I know right now 😂. I'm still up for learning others though and was willing to learn Premake should it be the solution we wanted to. On the other hand, I'm willing to try to give wxWidgets' CMake build system some quality-of-life improvements if they help.

In thinking about it, it seems like the notion of generating build system files among most build systems seems to stem from the idea that such files are only meant to be strictly part of the build process. In other words, they are not meant to be maintained and modified by hand like you would in a regular, non-generated build system (or in a generated build system that's output is meant to be dealt with in such a way). Perhaps at this point, in the world of C/C++ build systems, the typical convention is to use a meta-build system, often CMake, Meson, or even autotools I dare say, that generates files for a build tool. It might just be that we have to deal with this way of doing things, but from _my_ perspective anyways, I feel that CMake is easy enough to use when building wxWidgets. I mean, I build wxWidgets via CMake and Ninja on Linux, and I would use CMake with Visual Studio on Windows. (I can't say about macOS because I don't have a Mac to test things on, but it'd likely be the same as Linux if not Xcode). In that regard, I'd argue that CMake follows the feel of autotools when building in that you create and `cd` into your build directory, run your `configure` script with your build options, and then you build via `make`. The only difference is that you must have CMake instead ahead of time unlike with autotools, but personally, I don't see this as an issue as it's a simple process to install CMake or you'll already have it installed. Heck you can even install it with Visual Studio via its installer too nowadays, but I've digressed quite significantly at this point, so anyways...

Finally, I have a bit of an argument for dropping the autotools build system (mainly because CMake takes its place), but that'd be out of scope for this issue. I will warn you, however, that I am _very_ sue-happy when it comes to cleaning things up and ripping things out, so if we proceed with dropping something, I will be up for doing so! 😄

**Comment by Brickmaster1** [comment]
To be completely honest, I like CMake out of these options too, because I use XMake as my main build system and it is extremely simple to include CMake built packages/projects in your project (ex. wxWidgets's cmake now). For my specific use case, being able to use cross compiler with it and have the package be one unified thing I can cross compile with is really nice, I don't have to deal with the MSVS or XCode crap, I can just build the exact same deterministic way on every platform, to any platform.

**Comment by generic-pers0n** [comment]
> For my specific use case, being able to use cross compiler with it and have the package be one unified thing I can cross compile with is really nice, I don't have to deal with the MSVS or XCode crap, I can just build the exact same deterministic way on every platform, to any platform.

I think this also highlights another good point about CMake: you don't have to deal with what otherwise is, as I argue, an entirely different build system. Sure, the projects may be simple enough, but I argue that it would be even simpler to deal with just a single build system that works across all platforms.

On another note, I do find Xmake pretty interesting, and I'll probably experiment with it myself! Anyways, I've digressed 😄

**Comment by vadz** [comment]
We're not going to be able to do it before 3.3.4/3.4.0, there is still no good solution and we'll just have to live with it for now.

The longer term solution seems to be to just make CMake the primary and, eventually, only build system :-(
